### DISCUSSION TITLE
How to detect a self collision?

### DISCUSSION THREAD

**Discussion by Robokan** [isaac-sim/IsaacLab#652] [Q&A]
I have a Go1 with self collision enabled. I want to differentiate between collisions with the ground and collisions with itself in order to punish it with a negative reward or early termination during training. Ideally I would like to know the magnitude of the collisions as well. Anyone know how to do it?

**Comment by shaoxiang** [comment]
You can try using contact sensor to detect different contact collisions and measure collision forces.
The tutorial is very simple, you need to repeatedly review the examples:
https://isaac-sim.github.io/IsaacLab/source/tutorials/04_sensors/add_sensors_on_robot.html
> Welcome to join QQ isaac-sim group for communication: **723139415**

**Reply by Robokan** [reply]
Thanks for the link. This is what I have initially tried and wasn't able to detect any collisions.

I used this config.
``
self_collision: ContactSensorCfg = ContactSensorCfg(
        prim_path="/World/envs/env_.*/Robot/.*_calf", history_length=3, update_period=0.005, track_air_time=False,
        filter_prim_paths_expr=["/World/envs/env_.*/Robot/.*_calf"]
    )
``
in my code:

``
self_collision_forces = self.self_collision.data.force_matrix_w
self_collision_net_forces = self.self_collision.data.net_forces_w_history
if (torch.any(torch.abs(self_collision_forces) > 0.0)):
       print("self_collision_forces=",self_collision_forces)
if (torch.any(torch.abs(self_collision_net_forces) > 0.0)):
       print("self_collision_forces=",self_collision_net_forces)
``
Neither print statements happen even when there are self collisions between the "calf" bodies on the Go1.

**Reply by Mayankm96** [reply]
Is the sensor made part of the interactive scene? Did you add it to the scene or ensure that you call `sensor.update(dt)`?

**Reply by shaoxiang** [reply]
I have also encountered situations where contact sensor data cannot be read or data is all 0. It took me two days to solve this problem.
You can refer to it: #571 
Be sure to instantiate `ContactSensor` in `_setup_scene` and pass it to `self.scene.sensors["contact_sensor"]`
![image](https://github.com/isaac-sim/IsaacLab/assets/9045298/a1cdcff5-489d-4fb4-959d-9a7b0fbb407a)

**Reply by Robokan** [reply]
I use:
`
self.self_collision = ContactSensor(self.cfg.self_collision)
`
`
self.scene.sensors["self_collision"] = self.self_collision
`
in my _setup_scene

and
`
self.scene.update(dt=self.physics_dt)
`
before I check:
`
self.self_collision.data.force_matrix_w
`

which always contains zeros.

Interestingly the force_matrix_w has a shape of (num_envs, 4, 1, 3). If there are 4 calf on the Go1 why isn't this matrix (num_env,4,4,3)?

From what I can tell looking at contact_sensor.py line: 260 
self._contact_physx_view decides it has only one filter because filter_prim_paths_expr=["/World/envs/env_.*/Robot/.*_calf"] in my config is just one entry even though it should resolve to 4 filtered bodies. Then on line: 296 it uses that 1 filter to create self._data.force_matrix_w. The result is the shape: (num_envs, num_bodies, num_filters, 3). If we are looking at collisions between 4 bodies and 4 filtered bodies then num_filters should be 4 not 1 shouldn't it?

Alternatively if I look at net_forces_w I do get forces when 'calf' collide with the other 'calf' but I also get collisions with the ground which is not in the filter.
